### PR TITLE
CATROID-291: Refactoring

### PR DESCRIPTION
**Pull request by ThomasSchwengler**
Refactoring for CATROID-291, fixed and reopened

* Create a `EnumUtils` helper class to avoid code duplications among enums.
* Simplify readability of `NumberFormatsTest` by fixating the expected output
* `FormulaElement`
  * Move access to static variables from within functions into parameters
  * Move common functions into a `common` package to reduce code duplication
    and enhance readability
  * Add a `function` package and move all function interpretation into classes following the open-closed principle
  * Split up many functions to improve readability
  * Simplify many functions
* `Formula`
  * Simplify many functions by reducing the number of concerns per function
* `CollisionDetection`
  * Improve memory consumption by reducing the number of memory allocations in loops
* Hide direct references to android context in wrappers
* Rename functions to be more precise
* Replace some class instances with mocks in tests where possible

*Please enter a short description of your pull request and add a reference to the Jira ticket.*

### Your checklist for this pull request
Please review the [contributing guidelines](https://github.com/Catrobat/Catroid/blob/develop/README.md) and [wiki pages](https://github.com/Catrobat/Catroid/wiki/) of this repository.

- [x] Include the name of the Jira ticket in the PR’s title
- [x] Include a summary of the changes plus the relevant context
- [x] Choose the proper base branch (*develop*)
- [x] Confirm that the changes follow the project’s coding guidelines
- [x] Verify that the changes generate no compiler or linter warnings
- [x] Perform a self-review of the changes
- [-] Verify to commit no other files than the intentionally changed ones
- [-] Include reasonable and readable tests verifying the added or changed behavior
- [x] Confirm that new and existing unit tests pass locally
- [x] Check that the commits’ message style matches the [project’s guideline](https://github.com/Catrobat/Catroid/wiki/Commit-Message-Guidelines)
- [x] Stick to the project’s gitflow workflow
- [x] Verify that your changes do not have any conflicts with the base branch
- [x] After the PR, verify that all CI checks have passed
- [ ] Post a message in the *catroid-stage* or *catroid-ide* [Slack channel](https://catrobat.slack.com) and ask for a code reviewer

### CODE REVIEW

File: catroid/src/main/java/org/catrobat/catroid/utils/NumberFormats.java [hunk 1]
```diff
@@ -40,15 +40,14 @@ public static String stringWithoutTrailingZero(String value) {
...
-			return Integer.toString(number);
-		} else if (number < 10_000) {
-			return Integer.toString(number / 1_000) + (number % 1_000 > 100 ? "."
-					+ Integer.toString((number % 1_000) / 100) : "") + "k";
-		} else if (number < 1_000_000) {
-			return Integer.toString(number / 1_000) + "k";
+	public static String toMetricPostfixNotation(int number) {
+		if (number >= 1_000_000) {
+			return number / 1_000_000 + "M";
+		} else if (number >= 10_000) {
```
Comment (gPathpp) [inline catroid/src/main/java/org/catrobat/catroid/utils/NumberFormats.java:46]: Unfortunately, this has not the same logic...
Prev:
1900 => 1.9k
10900 => 10k
New:
1900 => 1k
10900 => 10.9k

PS: Kotlin would be very nice in this case 😄

File: catroid/src/main/java/org/catrobat/catroid/utils/NumberFormats.java [hunk 2]
```diff
@@ -40,15 +40,14 @@ public static String stringWithoutTrailingZero(String value) {
...
 
-	public static String humanFriendlyFormattedShortNumber(final int number) {
-		if (number < 1_000) {
-			return Integer.toString(number);
-		} else if (number < 10_000) {
-			return Integer.toString(number / 1_000) + (number % 1_000 > 100 ? "."
-					+ Integer.toString((number % 1_000) / 100) : "") + "k";
-		} else if (number < 1_000_000) {
-			return Integer.toString(number / 1_000) + "k";
+	public static String toMetricPostfixNotation(int number) {
```
Comment (gPathpp) [inline catroid/src/main/java/org/catrobat/catroid/utils/NumberFormats.java:43]: I think toMetricUnitRepresentation /Notation would be a better name since the Postfix notation is associated with reverse polish notation and calculators.

File: catroid/src/test/java/org/catrobat/catroid/test/formulaeditor/FormulaTest.java [hunk 3]
```diff
@@ -49,19 +53,23 @@
...
+@RunWith(MockitoJUnitRunner.class)
 public class FormulaTest {
 
 	private List<InternToken> internTokenList;
 
 	@Before
 	public void setUp() {
-		internTokenList = new LinkedList<InternToken>();
+		internTokenList = new LinkedList<>();
+		if (ProjectManager.getInstance() == null) {
```
Comment (gPathpp) [inline catroid/src/test/java/org/catrobat/catroid/test/formulaeditor/FormulaTest.java:66]: PLEASE-EXPLAIN: Why is this needed?

File: catroid/src/test/java/org/catrobat/catroid/test/formulaeditor/parser/ParserTest.java [hunk 4]
```diff
@@ -43,33 +42,11 @@
...
-	private static final float LOOK_BRIGHTNESS = 70f;
-	private static final float LOOK_COLOR = 0f;
-	private static final float LOOK_SCALE = 90.3f;
-	private static final float LOOK_ROTATION = 30.7f;
-	private static final int LOOK_ZPOSITION = 3;
+	@Mock
 	private Sprite testSprite;
 
-	@Before
-	public void setUp() {
```
Comment (gPathpp) [inline catroid/src/test/java/org/catrobat/catroid/test/formulaeditor/parser/ParserTest.java:60]: Nice!!

File: catroid/src/main/java/org/catrobat/catroid/sensing/CollisionDetection.java [hunk 5]
```diff
@@ -44,19 +43,17 @@
 	private CollisionDetection() {
 	}
 
-	public static double checkCollisionBetweenLooks(Look firstLook, Look secondLook) {
+	public static boolean checkCollisionBetweenLooks(Look firstLook, Look secondLook) {
```
Comment (gPathpp) [inline catroid/src/main/java/org/catrobat/catroid/sensing/CollisionDetection.java:46]: Nice! 
Can you please also rename the function? areLookColliding(..) or smth similar. 
(In my opinion, look.collidesWith(look) and then the polygon check, would be the best)
But iff you want it to keep it that way it's also ok 👍

File: catroid/src/main/java/org/catrobat/catroid/ui/dialogs/FormulaEditorComputeDialog.java [hunk 6]
```diff
@@ -145,22 +149,40 @@ public void onAccuracyChanged(Sensor sensor, int accuracy) {
...
+			int height = computeTextView.getLineCount() * computeTextView.getLineHeight();
+			int heightMargin = (int) (height * 0.5);
+			params.width = ViewGroup.LayoutParams.MATCH_PARENT;
+			params.height = height + heightMargin;
+			computeTextView.setLayoutParams(params);
 		});
 	}
+
+	private static class AndroidStringProvider implements Formula.StringProvider {
+		private final Context context;
```
[the remark this answers is not in the document]
Comment (ThomasSchwengler) [reply catroid/src/main/java/org/catrobat/catroid/ui/dialogs/FormulaEditorComputeDialog.java:172]: Good Idea!

File: catroid/src/main/java/org/catrobat/catroid/utils/NumberFormats.java (hunk 1 above)
Comment (ThomasSchwengler) [reply catroid/src/main/java/org/catrobat/catroid/utils/NumberFormats.java:46]: Oops, I must have missed that, will fix it (and probably convert it to kotlin)

File: catroid/src/main/java/org/catrobat/catroid/utils/NumberFormats.java (hunk 2 above)
Comment (ThomasSchwengler) [reply catroid/src/main/java/org/catrobat/catroid/utils/NumberFormats.java:43]: Done, thanks!

File: catroid/src/main/java/org/catrobat/catroid/sensing/CollisionDetection.java (hunk 5 above)
Comment (ThomasSchwengler) [reply catroid/src/main/java/org/catrobat/catroid/sensing/CollisionDetection.java:46]: Hmm. Should I then rename all other functions in this class as well? Maybe Kotlin will help here with some extension functions

File: catroid/src/main/java/org/catrobat/catroid/ui/ScratchProgramDetailsActivity.java [hunk 7]
```diff
@@ -270,11 +269,11 @@ private void onProgramDataUpdated() {
 		}
 
 		((TextView) findViewById(R.id.scratch_project_favorites_text))
-				.setText(humanFriendlyFormattedShortNumber(programData.getFavorites()));
```
[the remark this answers is not in the document]
Comment (ThomasSchwengler) [reply catroid/src/main/java/org/catrobat/catroid/ui/ScratchProgramDetailsActivity.java:273]: Ikr! :joy:

File: catroid/src/main/java/org/catrobat/catroid/utils/NumberFormats.java (hunk 1 above)
Comment (ThomasSchwengler) [reply catroid/src/main/java/org/catrobat/catroid/utils/NumberFormats.java:46]: Done

File: catroid/src/test/java/org/catrobat/catroid/test/formulaeditor/FormulaTest.java (hunk 3 above)
Comment (ThomasSchwengler) [reply catroid/src/test/java/org/catrobat/catroid/test/formulaeditor/FormulaTest.java:66]: The project manager instance might be null - this is not checked everywhere. Might be a good moment to remove the need for it in the test. I will see what I can do :joy:

File: catroid/src/test/java/org/catrobat/catroid/test/formulaeditor/parser/ParserTest.java (hunk 4 above)
Comment (ThomasSchwengler) [reply catroid/src/test/java/org/catrobat/catroid/test/formulaeditor/parser/ParserTest.java:60]: Thanks!

File: catroid/src/main/java/org/catrobat/catroid/sensing/CollisionDetection.java (hunk 5 above)
Comment (gPathpp) [reply catroid/src/main/java/org/catrobat/catroid/sensing/CollisionDetection.java:46]: It's just more work, so you don't have to. 
I think extension functions are not necessary because we have access to class "Look" :smile:

**Comment by DinosaurierRex** [comment]
Rebased and reopend -> [PR-3642](https://github.com/Catrobat/Catroid/pull/3642)
